### PR TITLE
Add pinksale adapter

### PR DESCRIPTION
**Pull request by pinkmoonfinance**
##### Twitter Link:
https://twitter.com/pinkecosystem

##### List of audit links if any:
https://www.certik.com/projects/pinksale

##### Website Link:
https://pinksale.finance/

##### Logo (High resolution, preferably in .svg and .png, for application on both white and black backgrounds. Will be shown with rounded borders):
https://www.pinksale.finance/logo512.png

##### Current TVL:
$399.7M

##### Chain:
BSC

##### Coingecko ID (so your TVL can appear on Coingecko): (https://api.coingecko.com/api/v3/coins/list)
pinksale

##### Coinmarketcap ID (so your TVL can appear on Coinmarketcap): (https://api.coinmarketcap.com/data-api/v3/map/all?listing_status=active,inactive,untracked&start=1&limit=10000)
pinksale

##### Short Description (to be shown on DefiLlama):
PinkSale helps everyone to create their own tokens and token sales in few seconds. Tokens created on PinkSale will be verified and published on explorer websites.

##### Token address and ticker if any:
0x602ba546a7b06e0fc7f58fd27eb6996ecc824689

##### Category (full list at https://defillama.com/categories) *Please choose only one:
Launchpad

##### forkedFrom (Does your project originate from another project):
No

##### methodology (what is being counted as tvl, how is tvl being calculated):

TVL explain

All assets locked in the Pinklock contract will be counted as TVL. This can be seen at https://www.pinksale.finance/#/pinklock/liquidity and https://www.pinksale.finance/#/pinklock/token
Pinklock address:

BSC: 0x7ee058420e5937496F5a2096f04caA7721cF70cc

And then, we get all amount of each asset locked in the contract. 
Finally, we calculate the price by the amount and show it as TVL.

### CODE REVIEW

**Comment by llamatester** [comment]
The adapter at projects/pinksale/index.js exports TVL: 
        
 
 
    avalanche                 0
    polygon                   0
    cronos                    0
    fantom                    0
    kcc                       0
    ethereum                  0
    bsc                       399.92 M
    
    total                    399.92 M

**Comment by waynebruce0x** [comment]
Can we calculate this using onchain calls instead please? Can any TVL from PINKSALE tokens and LPs be exported as staking and pool2 respectively?

**Comment by pinkmoonfinance** [comment]
> Can we calculate this using onchain calls instead please? Can any TVL from PINKSALE tokens and LPs be exported as staking and pool2 respectively?

ok

**Comment by llamatester** [comment]
The adapter at projects/pinksale exports TVL: 
        
 
 
    kcc                       0
    cronos                    29.86 k
    ethereum                  9.59 M
    avalanche                 924.2907478368639
    polygon                   25.54 k
    fantom                    1.05 k
    bsc                       297.93 M
    
    total                    307.57 M

**Comment by pinkmoonfinance** [comment]
Hey @waynebruce0x 
Here is our API for TVL: https://github.com/pinkmoonfinance/pinksale-api-tvl
Can you take a look?

**Comment by waynebruce0x** [comment]
How do tokens get into pinklock? Do teams deposit tokens for vesting etc?

**Comment by pinkmoonfinance** [comment]
Yes, teams deposit tokens and LP to lock on PinkLock. And every presale on PinkSale.Finance will lock LP on PinkLock as well.

File: projects/pinksale/index.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    polygon(),
+    kcc(),
+    fantom(),
+    avalanche(),
+    cronos(),
+  ]);
+  return bscTvl + ethTvl + polygonTvl + kccTvl + ftmTvl + avaxTvl + croTvl;
+}
+
+fetch().then((tvl) => console.log(tvl))
```
Comment (g1nt0ki) [inline projects/pinksale/index.js:61]: please remove this line

**Review by g1nt0ki — CHANGES_REQUESTED** [review]
can you compute the tvl using on chain calls instead of calls to your api?

**Comment by llamatester** [comment]
Error while running adapter at projects/pinksale: 
        
 
  
    
    
    Error: TVL for key tvl is not a number, instead it is NaN
    Truncated error stack:
    
        at getTvl (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:79:11)
        at async Promise.all (index 7)
        at async /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:205:3

**Comment by llamatester** [comment]
The adapter at projects/pinksale exports TVL: 
        
 
 
    polygon                   33.91 k
    fantom                    3.36 k
    bsc                       261.45 M
    ethereum                  9.01 M
    cronos                    21.38 k
    avalanche                 719.5333017698808
    
    total                    270.53 M

**Comment by pinkmoonfinance** [comment]
> can you compute the tvl using on chain calls instead of calls to your api?

We can't calculate TVL on chan. Because of this task is very heavy with RPC by each chain, it may be timeout. So we need database to calculate row by row to store TVL. We also update once per day for TVL change.

**Comment by g1nt0ki** [comment]
> > can you compute the tvl using on chain calls instead of calls to your api?
> 
> We can't calculate TVL on chain. Because of this task is very heavy with RPC by each chain, it may be timeout. So we need database to calculate row by row to store TVL. We also update once per day for TVL change.

Hi, sorry for the delay, I am trying to calc tvl on chain, you can look at the code here: https://github.com/DefiLlama/DefiLlama-Adapters/pull/2882 

you can follow the development there or else, I can post here once it is ready

**Comment by g1nt0ki** [comment]
BSC vault address is `0x7ee058420e5937496F5a2096f04caA7721cF70cc` , can you please provide the vault addresses in other chains?

**Comment by pinkmoonfinance** [comment]
> > > can you compute the tvl using on chain calls instead of calls to your api?
> > 
> > 
> > We can't calculate TVL on chain. Because of this task is very heavy with RPC by each chain, it may be timeout. So we need database to calculate row by row to store TVL. We also update once per day for TVL change.
> 
> Hi, sorry for the delay, I am trying to calc tvl on chain, you can look at the code here: #2882
> 
> you can follow the development there or else, I can post here once it is ready

hi, thanks for that, we have 2 version for pinklock now, I just sent to you

**Pull request closed without merging**